pkg/api/grpc_server.go

```diff
@@ -0,0 +1,204 @@
+			return nil, handleFulcioGRPCError(ctx, codes.InvalidArgument, err, invalidPublicKey)
+		}
+	}
+
+	// Validate public key, checking for weak key parameters.
+	if err := cryptoutils.ValidatePubKey(publicKey); err != nil {
+		return nil, handleFulcioGRPCError(ctx, codes.InvalidArgument, err, insecurePublicKey)
+	}
+
+	subject, err := challenges.ExtractSubject(ctx, principal, publicKey, request.ProofOfPossession)
+	if err != nil {
+		return nil, handleFulcioGRPCError(ctx, codes.InvalidArgument, err, invalidSignature)
+	}
+
+	var csc *certauth.CodeSigningCertificate
+	var sctBytes []byte
+	// TODO: prefer embedding SCT if possible
+	if _, ok := g.ca.(certauth.EmbeddedSCTCA); !ok {
+		// currently configured CA doesn't support pre-certificate flow required to embed SCT in final certificate
+		csc, err = g.ca.CreateCertificate(ctx, subject)
+		if err != nil {
+			// if the error was due to invalid input in the request, return HTTP 400
+			if _, ok := err.(certauth.ValidationError); ok {
+				return nil, handleFulcioGRPCError(ctx, codes.InvalidArgument, err, err.Error())
+			}
+			// otherwise return a 500 error to reflect that it is a transient server issue that the client can't resolve
+			return nil, handleFulcioGRPCError(ctx, codes.Internal, err, genericCAError)
+		}
+
+		// Submit to CTL
+		if g.ct != nil {
+			sct, err := g.ct.AddChain(ctx, csc)
+			if err != nil {
+				return nil, handleFulcioGRPCError(ctx, codes.Internal, err, failedToEnterCertInCTL)
+			}
+			sctBytes, err = json.Marshal(sct)
+			if err != nil {
+				return nil, handleFulcioGRPCError(ctx, codes.Internal, err, failedToMarshalSCT)
+			}
+		} else {
+			logger.Info("Skipping CT log upload.")
+		}
+	}
+
+	metricNewEntries.Inc()
+
+	finalPEM, err := csc.CertPEM()
+	if err != nil {
+		return nil, handleFulcioGRPCError(ctx, codes.Internal, err, failedToMarshalCert)
+	}
+
+	finalChainPEM, err := csc.ChainPEM()
+	if err != nil {
+		return nil, handleFulcioGRPCError(ctx, codes.Internal, err, failedToMarshalCert)
+	}
+
+	result := &fulciogrpc.SigningCertificate{
+		Chain: &fulciogrpc.CertificateChain{
+			Certificates: []string{strings.TrimSpace(string(finalPEM)), strings.TrimSpace(string(finalChainPEM))},
+		},
+	}
+
+	if len(sctBytes) > 0 {
+		result.SignedCertificateTimestamp = sctBytes
+	}
+
+	return result, nil
+}
+
+func (g *grpcCAServer) GetTrustBundle(ctx context.Context, _ *empty.Empty) (*fulciogrpc.TrustBundle, error) {
+	logger := log.ContextLogger(ctx)
+
+	root, err := g.ca.Root(ctx)
+	if err != nil {
+		logger.Error("Error retrieving root cert: ", err)
+		return nil, handleFulcioGRPCError(ctx, codes.Internal, err, genericCAError)
+	}
+
+	return &fulciogrpc.TrustBundle{
+		Chains: []*fulciogrpc.CertificateChain{{
+			Certificates: []string{string(root)},
+		}},
+	}, nil
+}
+
+func extractIssuer(token string) (string, error) {
+	parts := strings.Split(token, ".")
+	if len(parts) < 2 {
+		return "", fmt.Errorf("oidc: malformed jwt, expected 3 parts got %d", len(parts))
```

Comment:
Should we enforce that parts == 3, not just >= 2?

Reply:
That was a copy/paste of code from `ca.go` - I agree with the proposed change though.